qmi8658c/qmi8658c.go

```diff
@@ -0,0 +1,186 @@
+		AccLowPass:  ACC_LOW_PASS_2_62,
+		GyroEnable:  GYRO_FULL_ENABLE,
+		GyroScale:   GYRO_512DPS,
+		GyroRate:    GYRO_1000HZ,
+		GyroLowPass: GYRO_LOW_PASS_2_62,
+	}
+}
+
+// Check if the usar has defined a desired configuration, if not uses the
+// DefaultConfig, then defines the AccLsbDiv and GyroLsbDiv based on the
+// configurations and, finally, send the commands and configure the IMU.
+func (d *Device) Configure(cfg Config) {
+	if cfg == (Config{}) {
+		cfg = DefaultConfig()
+	}
+	var val uint16
+	// Setting accelerometer LSB
+	switch cfg.AccScale {
+	case ACC_2G:
+		d.AccLsbDiv = 1 << 14
+	case ACC_4G:
+		d.AccLsbDiv = 1 << 13
+	case ACC_8G:
+		d.AccLsbDiv = 1 << 12
+	case ACC_16G:
+		d.AccLsbDiv = 1 << 11
+	default:
+		d.AccLsbDiv = 1 << 12
+	}
+	// Setting gyro LSB
+	switch cfg.GyroScale {
+	case GYRO_16DPS:
+		d.GyroLsbDiv = 2048
+	case GYRO_32DPS:
+		d.GyroLsbDiv = 1024
+	case GYRO_64DPS:
+		d.GyroLsbDiv = 512
+	case GYRO_128DPS:
+		d.GyroLsbDiv = 256
+	case GYRO_256DPS:
+		d.GyroLsbDiv = 128
+	case GYRO_512DPS:
+		d.GyroLsbDiv = 64
+	case GYRO_1024DPS:
+		d.GyroLsbDiv = 32
+	case GYRO_2048DPS:
+		d.GyroLsbDiv = 16
+	default:
+		d.GyroLsbDiv = 64
+	}
+	// SPI Modes
+	val = uint16((cfg.SPIMode | cfg.SPIEndian | cfg.SPIAutoInc))
+	d.WriteRegister(CTRL1, val)
+	// Accelerometer config
+	val = uint16(cfg.AccScale | cfg.AccRate)
+	d.WriteRegister(CTRL2, val)
+	// Gyro config
+	val = uint16(cfg.GyroScale | cfg.GyroRate)
+	d.WriteRegister(CTRL3, val)
+	// Sensor DSP config
+	val = uint16(cfg.GyroLowPass | cfg.AccLowPass)
+	d.WriteRegister(CTRL5, val)
+	// Sensors config
+	val = uint16(cfg.GyroEnable | cfg.AccEnable)
+	d.WriteRegister(CTRL7, val)
+}
+
+// Read the acceleration from the sensor, the values returned are in mg
+// (milli gravity), which means that 1000 = 1g.
+func (d *Device) ReadAcceleration() (x int32, y int32, z int32) {
+	data := make([]byte, 6)
+	raw := make([]int32, 3)
+	d.ReadRegister(ACC_XOUT_L, data)
+	for i := range raw {
+		raw[i] = int32(uint16(data[(2*i+1)])<<8 | uint16(data[i]))
+		if raw[i] >= 32767 {
+			raw[i] = raw[i] - 65535
+		}
+	}
+	x = -raw[0] * 1000 / int32(d.AccLsbDiv)
+	y = -raw[1] * 1000 / int32(d.AccLsbDiv)
+	z = -raw[2] * 1000 / int32(d.AccLsbDiv)
+	return x, y, z
+}
+
+// Read the rotation from the sensor, the values returned are in mdeg/sec
+// (milli degress/second), which means that a full rotation is 360000.
+func (d *Device) ReadRotation() (x int32, y int32, z int32) {
+	data := make([]byte, 6)
+	raw := make([]int32, 3)
+	d.ReadRegister(GYRO_XOUT_L, data)
+	for i := range raw {
+		raw[i] = int32(uint16(data[(2*i+1)])<<8 | uint16(data[i]))
+		if raw[i] >= 32767 {
+			raw[i] = raw[i] - 65535
+		}
+	}
+	x = raw[0] * 1000 / int32(d.GyroLsbDiv)
+	y = raw[1] * 1000 / int32(d.GyroLsbDiv)
+	z = raw[2] * 1000 / int32(d.GyroLsbDiv)
+	return x, y, z
+}
+
+// Read the temperature from the sensor, the values returned are in Celsius.
+func (d *Device) ReadTemperature() (t float32) {
```

Comment:
Please, follow the signature  `ReadTemperature() (int32, error) {` as other drivers do. It should return millidegrees:

https://github.com/tinygo-org/drivers/blob/release/bme280/bme280.go#L116


As a rule of thumb, avoid float when possible.

Reply:
Changed the signature as proposed and updated the example file.

My main language is Python and I'm new to Go and TinyGo, so I have to ask, this would be for memory optimization? Since float32 would use a lot more memory than int32?

Reply:
I don't know the details, but yes, the boards targeted don't have much memory and don't play well with floats in general.